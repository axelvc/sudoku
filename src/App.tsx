import { FC } from 'react'
import { Provider } from 'react-redux'

import Controls from './controls/Controls'
import store from './store/store'
import Sudoku from './sudoku/Sudoku'

const App: FC = () => (
  <main>
    <Provider store={store}>
      <Sudoku />
      <Controls />
    </Provider>
  </main>
)

export default App
