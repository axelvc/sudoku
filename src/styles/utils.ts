import { css } from 'styled-components'

export const breakpoints = {
  md: '768px',
}

export const roundedGrid = (rows: number, cols: number): ReturnType<typeof css> => css`
  display: grid;
  grid-template: ${p => `repeat(${rows}, 1fr) / repeat(${cols}, 1fr) `};
  gap: 0.25rem;
  overflow: hidden;
  border-radius: 0.5rem;
`

export const fancyBox = css`
  padding: 1.5rem;
  background: var(--color-gray-100);
  border-radius: 1rem;
`
