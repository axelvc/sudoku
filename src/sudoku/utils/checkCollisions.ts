import { Coords, BoxData, SudokuState } from '../sudokuSlice'

function getRowSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle[row].filter((box, i) => box.value && i !== col)
}

function getColSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle.reduce((siblings, r, i) => {
    const box = r[col]

    if (box.value && i !== row) {
      siblings.push(box)
    }

    return siblings
  }, [])
}

function getBlockSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  const blockRow = Math.floor(row / 3)
  const blockCol = Math.floor(col / 3)

  return puzzle.reduce((siblings, r, rI) => {
    r.forEach((box, cI) => {
      if (
        box.value &&
        rI !== row &&
        cI !== col &&
        blockRow === Math.floor(rI / 3) &&
        blockCol === Math.floor(cI / 3)
      ) {
        siblings.push(box)
      }
    })

    return siblings
  }, [])
}

export default function checkCollisions(
  state: SudokuState,
  { row, col }: Coords,
  oldValue: number,
): void {
  const { puzzle } = state
  const box = puzzle[row][col]
  const siblings = [
    ...getRowSiblings(puzzle, { row, col }),
    ...getColSiblings(puzzle, { row, col }),
    ...getBlockSiblings(puzzle, { row, col }),
  ]

  if (box.errors) {
    box.errors = 0

    siblings.forEach(sibling => {
      if (sibling.value === oldValue) {
        sibling.errors--
      }
    })
  }

  siblings.forEach(sibling => {
    if (sibling.value === box.value) {
      sibling.errors++
      box.errors++
    }
  })
}
