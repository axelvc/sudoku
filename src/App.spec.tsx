import { render, screen, userEvent, regex } from './test/utils'
import App from './App'

jest.mock('./sudoku/sudokuService')

beforeEach(() => render(<App />))

describe('integration between controls and sudoku', () => {
  describe('numpad', () => {
    const numpadNumbers = [1, 2, 3, 4, 5, 6, 7, 8, 9].map(String)

    it('should fill box value', () => {
      const $box = screen.getByTestId('box-0-0')

      numpadNumbers.forEach(n => {
        const $control = screen.getByTestId(`numpad-${n}`)

        userEvent.click($control)
        userEvent.click($box)

        expect($box).toHaveTextContent(n)
      })
    })

    it('should add mark value', () => {
      const $box = screen.getByTestId('box-0-0')
      const $marks = screen.getByRole('button', { name: regex('marks') })

      userEvent.click($marks)

      numpadNumbers.forEach(n => {
        const $control = screen.getByTestId(`numpad-${n}`)

        userEvent.click($control)
        userEvent.click($box)
      })

      expect($box).toHaveTextContent(numpadNumbers.join(''))
    })
  })

  describe('utils controls', () => {
    it('should clean box on click with "erase" control enabled', () => {
      const $box = screen.getByTestId('box-0-0')
      const $erase = screen.getByRole('button', { name: regex('erase') })
      const $numpadControl = screen.getByTestId('numpad-1')

      userEvent.click($numpadControl)
      userEvent.click($box)
      userEvent.click($erase)
      userEvent.click($box)

      expect($box).toHaveTextContent('')
    })

    it('should add error styles on click "validate" control', () => {
      const $validate = screen.getByRole('button', { name: regex('validate') })
      const $box = screen.getByTestId('box-0-0')
      const $numpad1 = screen.getByTestId('numpad-1')

      userEvent.click($numpad1)
      userEvent.click($box)
      userEvent.click($validate)

      expect($box).toHaveStyleRule('--color', 'var(--color-red-500)')
      expect($box).toHaveStyleRule('--background', 'var(--color-red-100)')
      expect($box).toMatchSnapshot()
    })

    it('should restore value on click "undo" control after set value', () => {
      const $undo = screen.getByRole('button', { name: regex('undo') })
      const $box = screen.getByTestId('box-0-0')
      const $numpad1 = screen.getByTestId('numpad-1')

      userEvent.click($numpad1)
      userEvent.click($box)
      userEvent.click($undo)

      expect($box).toHaveTextContent('')
    })

    it('should restore all changed values on click "reset" control', () => {
      const $reset = screen.getByRole('button', { name: regex('reset') })
      const $box1 = screen.getByTestId('box-0-0')
      const $box2 = screen.getByTestId('box-0-1')
      const $numpad1 = screen.getByTestId('numpad-1')

      userEvent.click($numpad1)
      userEvent.click($box1)
      userEvent.click($box2)

      userEvent.click($reset)
      expect($box1).toHaveTextContent('')
      expect($box2).toHaveTextContent('')
    })
  })
})
