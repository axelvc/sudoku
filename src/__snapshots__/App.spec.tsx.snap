// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration between controls and sudoku utils controls should add error styles on click "validate" control 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: column;
  gap: 0.25rem;
  place-items: center;
  min-height: 32px;
  padding: 0 0.75rem;
  font-weight: 600;
  color: var(--color);
  text-transform: capitalize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: var(--background);
  border-radius: 0.5rem;
  outline-color: var(--color);
  -webkit-transition: 100ms ease-out;
  transition: 100ms ease-out;
  -webkit-transition-property: background,color;
  transition-property: background,color;
  --color: var(--color-blue-400);
  --color-active: var(--color-gray-100);
  --background: var(--color-blue-100);
  --background-hover: var(--color-blue-200);
  --background-active: var(--color-blue-400);
}

.c0:hover,
.c0:focus-visible {
  background: var(--background-hover);
}

.c0:active {
  color: var(--color-active);
  background: var(--background-active);
}

.c1 {
  position: relative;
  aspect-ratio: 1;
  display: grid;
  place-items: center;
  font-weight: 700;
  border-radius: 0.2rem;
  --color: var(--color-gray-900);
  --color-active: var(--color-gray-900);
  --background: var(--color-gray-200);
  --background-hover: var(--color-gray-300);
  --background-active: var(--color-gray-400);
  --disabled-line: var(--color-gray-300);
  --color: var(--color-red-500);
  --background: var(--color-red-100);
  --background-hover: var(--color-red-300);
  --color-active: var(--color-gray-100);
  --background-active: var(--color-red-500);
  --disabled-line: var(--color-red-300);
}

.c1:disabled {
  background-image: linear-gradient( -45deg, var(--background) 15%, var(--disabled-line) 15% 25%, var(--background) 25% 35%, var(--disabled-line) 35% 45%, var(--background) 45% 55%, var(--disabled-line) 55% 65%, var(--background) 65% 75%, var(--disabled-line) 75% 85%, var(--background) 85% );
}

<button
  class="c0 c1"
  data-testid="box-0-0"
>
  <span>
    1
  </span>
</button>
`;
